supernova portaudio: fix for requesting number of I/O channels exceeding hardware capabilities

dyfer opened the pull request:
<!-- Please see CONTRIBUTING.md for guidelines. -->

## Purpose and Motivation

Fixes #4029 

~~This is WIP - it works as expected, but the implementation is hack-y.~~

The gist of the issue was that in ```void nova_server::prepare_backend(void)```, as well as in ```inline void run_scheduler_tick(void)``` (in `server.hpp`/`server.cpp`) we needed to use number of input channels clipped to the number of available ins in the hardware, but number of output channels as requested by the user (_not_ clipped to what HW provides).

After modifications to `server.cpp` and `server.hpp` and making `input_channels`/`output_channels` variables available globally to fix this, I have also tested JACK and non-real-time (soundfile) backends; both seem to work as expected. (Soundfile needed a fix, which is now included).

**I'm open to suggestions if there's a better way to fix this :)**

BTW original issue was reported only for number of output channels, but it actually applies to exceeding both input and output HW channels. This PR contains fixes both. To test:

```supercollider
(
Server.supernova;
s.options.numOutputBusChannels_(4); //choose number larger than the soundcard provides
s.options.numInputBusChannels_(4); //choose number larger than the soundcard provides
s.reboot;
)
s.meter; 

x.free; x = {PinkNoise.ar(-12.dbamp)}.play(outbus: 4, target: 0, addAction: \addToHead); //play on the first input channel

x.free; x = {PinkNoise.ar(-12.dbamp)}.play(outbus: 6, target: 0, addAction: \addToHead); //play on the input channel above what hardware provides

x.free;
```

## Types of changes

<!-- Delete lines that don't apply -->

- Bug fix

## To-do list

<!-- Complete an item by checking it: [x]. Add new entries to track your progress -->

- [x] Code is tested
  - [x] PortAudio input/output is tested
  - [x] JACK input/output is tested
  - [x] non-real-time (soundfile) input/output is tested
- [x] Code structure/logic is improved
- [ ] Commit history is cleaned up (after review)
- [x] This PR is ready for review

dyfer commented:
Hey @sonoro1234 could you possibly test this on Windows?

sonoro1234 commented:
Tested and working!!

brianlheim commented:
this is next on my list but it's going to take some time to understand and review the code.

@sonoro1234 -- in the future, please *do not* approve PRs unless you think they are ready for merge. you can just leave a comment that you've tested the changes. (see https://github.com/supercollider/supercollider/wiki/%5BWIP%5D-Reviewer-instructions#using-reviews)

brianlheim commented:
question -- what does scsynth do in this scenario?

dyfer commented:
> question @dyfer -- what does scsynth do in this scenario?

scsynth behaves like supernova after this PR - one can specify any number of I/O channels. They can be used as inputs/outputs in the synth, but only the first "x" will be connected to the hardware (as available).

A typical use case:
A live-processing multi-input project is tested without the multichannel soundcard. Number of inputs are configured, so it would break if the number of input channels changes - when testing with the built-in soundcard, one needs to be able to specify extra input channels for the patch to work. Moreover, if one has a mutlichannel recording of the input channels signal for testing, one can play that at the top of the root node to the input channels and the patch can be tested that way.

brianlheim commented:
thanks! going to take me some time to review this. i'm not very familiar with the surrounding code.

dyfer commented:
I rebased this PR to resolve conflicts with the current `develop`.

brianlheim commented on server/supernova/server/server.hpp:
```diff
@@ -226,18 +228,16 @@ class nova_server : public asynchronous_log_thread,
 inline void run_scheduler_tick(void) {
     const int blocksize = sc_factory->world.mBufLength;
-    const int input_channels = sc_factory->world.mNumInputs;
-    const int output_channels = sc_factory->world.mNumOutputs;
```
to me it's a bad sign that the server and sc_factory (which is a unique_ptr to SC_World) have different ideas of the number of ins and outs. can you delay instantiating sc_factory until that's known, in nova_server::prepare_backend()? or otherwise, perhaps modify these values during audio backend config?

dyfer commented:
Thanks so much @brianlheim for the review and observations. I'm looking into this but it will take few days before I can really dive into this again. BTW I resolved conflicts locally but I'm not pushing them until I look more into addressing your comments.

dyfer commented:
> this doesn't quite seem right to me... you've added 2 new members to server called input_channels and output_channels, but the server already has 2 members of those names that it inherits from audio_backend, via audio_settings_basic. 

Thanks for pointing it out! Now I see that I could've access the correct number of channels through audio_backend there, which I'm now doing.

> next, i don't see what differentiates portaudio's mInputChannelCount from the server's input_channels, don't they end up being the same value after open_stream? do you need that member? 

You're right, only number of output channels becomes different. But wouldn't it be weird if we used a pair of `mHwOutputChannels` and `input_channels` throughout the class? I thought it is more clear to members for both input and output channels numbers.

> also, if you renamed those portaudio members to something like "mHwOutputChannels" (...)

Good point, updated!

How does it look now? Do I need to think about server and sc_factory having a different idea about number of ins/outs?

brianlheim commented on server/supernova/audio_backend/portaudio_backend.hpp:
```diff
@@ -69,6 +69,11 @@ class portaudio_backend : public detail::audio_backend_base<sample_type, float,
+private:
+    // these values will be clipped to the number of available hardware ins/outs
+    uint16_t mHwInputChannels = 0;
+    uint16_t mHwOutputChannels = 0;
```
remove `mHwOutputChannels`

brianlheim commented on server/supernova/audio_backend/portaudio_backend.hpp:
```diff
@@ -69,6 +69,11 @@ class portaudio_backend : public detail::audio_backend_base<sample_type, float,
 
+private:
+    // these values will be clipped to the number of available hardware ins/outs
+    uint16_t mHwInputChannels = 0;
```
rename to `m_hwInputChannels` per [webkit naming conventions](https://webkit.org/code-style-guidelines/#names-data-members)

brianlheim commented:
i re-ran CI, seems OK now. sometimes there are just server outages, i guess

dyfer commented:
All right, I removed superfluous member. I also renamed variables and in `run_scheduler_tick` only changed `input_channels`. 

I tried to change that value in the initialization of sc_factory but couldn't figure it out. I added a FIXME note though.

How does this look?

(will rebase to cleanup later)

brianlheim commented on server/supernova/audio_backend/portaudio_backend.hpp:
```diff
@@ -202,10 +206,14 @@ class portaudio_backend : public detail::audio_backend_base<sample_type, float,
+        // input_channels is clipped to the number of available audio inputs (above)
+        // output_channels is equal to the requested number of outputs (not clipped)
+        // input_samples and output_samples vectors are resized
+        // to the number of available audio inputs/outputs
```
these last two lines are not really necessary, they just restate what the code is already doing.

brianlheim commented on server/supernova/server/server.hpp:
```diff
@@ -226,7 +226,11 @@ class nova_server : public asynchronous_log_thread,
+    // FIXME/TODO
+    // sc_factory and server end up with different values for number of input channels
+    // this was done in pr 4575 to fix PortAudio issue when requesting more channels than the harware provides
+    // consider making these values unified
```
i would recommend "server may have fewer input channels than sc_factory"; if i were reading this fresh i would want to know that, because it rules out the possibility of certain bugs (i.e., i don't have to go and check whether it's possible for server to have _more_ input channels, and i also don't have to check whether that would even cause problems).

PR # and reason are not particularly relevant here, in my opinion, and can be found with git blame/log. personally i would leave them out, but it's up to you.

dyfer replied:
That sounds good, I used your suggestion. However, do you think whether another comment needs to be added to prevent someone from just switching this back to `sc_factory->world.mNumInputs;`? The mistake won't be immediately apparent, unless they test it requesting higher number of channels than the hardware provides.

Sounds like a task for a unit test, but testing in server/cpp is uncharted territory for me.

brianlheim replied:
it looks good to me as is. i would hope anyone reviewing would be suspicious of that kind of change happening with a "FIXME" right above it :)

brianlheim commented:
> EDIT: I also realized that in portaudio_backend I could minimize the number of changes by just storing requested number of channels in outchans_requested or m_outchansRequested and then I wouldn't have to touch other places with output_channels... would that be better?

i'm not quite sure this is what you mean, but if you're proposing to make the PA backend hold the number of 'server' channels and the server hold the number of 'hardware' channels, that seems like an inversion of responsibilities...

dyfer commented:
> i'm not quite sure this is what you mean, but if you're proposing to make the PA backend hold the number of 'server' channels and the server hold the number of 'hardware' channels, that seems like an inversion of responsibilities...

That's not quite what I was proposing, but now I realized that my suggestion wouldn't really work. I think the current implementation is as clean as possible.

dyfer commented:
thanks! I also rebased this to squash commits.
